mbuild/utils/orderedset.py

```diff
@@ -0,0 +1,66 @@
+"""Ordered set module."""
+from collections.abc import MutableSet
+from copy import deepcopy
+
+
+class OrderedSet(MutableSet):
+    """An ordered set object with additional convenience methods.
+
+    Taken from code suggested by Vyas in
+    https://github.com/mosdef-hub/mbuild/issues/865
+
+    Methods
+    -------
+    add
+    discard
+    union
+    intersection
+    difference
+    """
+
+    def __init__(self, *args):
+        self._data = {value: None for value in args}
+
+    def __repr__(self):
+        """Return the OrderedSet representation."""
+        data = ", ".join([str(i) for i in self._data])
+        return f"{self.__class__.__name__}({data})"
+
+    def __contains__(self, key):
+        """Determine whether the element `key` is in the set."""
+        return key in self._data
+
+    def __getitem__(self, value):
+        """Get an item at index `value`."""
+        return list(self._data)[value]
```

Comment:
If you expect to create a set once, access its data frequently, but modify it infrequently, making a list every time could be a lot slower than creating one once and caching it. The catch is that you'd have to invalidate the cache every time the data structure is modified, so it really depends on how frequently you expect to modify. What you'd want to do is something like
```
    @property
    def _list(self):
        try:
            return self._cached_list
        except AttributeError:
            self._cached_list = list(self._data)
```
Then add `self.__dict__.pop('_cached_list', None)` to both `add` and `discard`. This will slow down both of those methods though (and any other methods that call them), so it's a question of whether you expect to mostly be doing read or write operations.

Reply:
> You may also want to override `clear`, `pop`, and `remove` in case you plan to use them frequently. Otherwise those will fall back to the default `MutableSet` implementation, which will be slower.

I've overwritten `remove` as it is commonly used. I don't think clear or pop are ever used.

Reply:
Looking at the code-base, add/ discard seem to be more frequently used than indexing the `OrderedSet`. Most read operations are happening through iteration, so I think this caching would be more expensive for us.

Reply:
I reverted the list caching (d09bf46). It might be worth looking into ways to profile mbuild and make it faster for large system, but that can be done in a later PR.